rfd/0159-password-status.md

```diff
@@ -66,6 +68,48 @@ ContactAdmin --> End([End])
 
 For clarity, error cases (such as password mismatch or inability to confirm user identity) were omitted from the above diagram.
 
+We are going to implement the following changes to the WebAuthn flow and APIs:
+
+1. Move the responsibility for verifying the old password to the `ChangePassword` RPC call. This will allow us to decouple the verification from setting the new password in the UI (the old password will not be required when generating the challenge). This verification will be specific to the `ChangePassword` call and will not be a part of the generic `LoginFlow`. The
+2. Add a possibility to request user verification through [`ChallengeExtensions`](https://github.com/gravitational/teleport/blob/3a8dcdc52c47b37091796fadb3ca775c06b7933e/api/proto/teleport/mfa/v1/mfa.proto#L27):
+   ```proto
+   message ChallengeExtensions {
+     // ...
+     string user_verification_requirement = 3; // "required", "discouraged", etc.
+   }
+   ```
+   This will automatically become a parameter of the `CreateAuthenticateChallenge` RPC call through the `ChallengeExtensions` field. To make it possible to specify from the client side, we will accept a corresponding `userVerificationRequirement` parameter to the `/v1/webapi/mfa/authenticatechallenge/password` POST endpoint.
+3. Create a dedicated [challenge scope](https://github.com/gravitational/teleport/blob/3a8dcdc52c47b37091796fadb3ca775c06b7933e/api/proto/teleport/mfa/v1/mfa.proto#L40) for the challenge that allows changing a password (`CHALLENGE_SCOPE_PASSWORD_CHANGE`).
+
+Here is what the updated process is going to look like:
+
+```mermaid
+sequenceDiagram
+participant C as Client
+participant P as Proxy Service
+participant A as Auth Service
+C ->> P: POST mfa/authenticatechallenge/password<br>Specify user verification requirement
+activate P
+P ->> A: CreateAuthenticateChallenge<br>Request: ContextUser<br>Specify user verification requirement<br>Set scope to PASSWORD_CHANGE
+deactivate P
+activate A
+A ->> P: Challenge
+deactivate A
+activate P
+P ->> C: Challenge
+deactivate P
+activate C
+C ->> P: PUT users/password<br>Send new pasword, assertion response<br>Send old password only if no UV requested
+deactivate C
+activate P
+P ->> A: ChangePassword
+deactivate P
+activate A
+A ->> A: Verify scope and assertion response
+A ->> A: Verify old password if UV=0
```

Comment:
Suggestion:

```suggestion
A ->> A: if old password is empty:<br/>UserVerificationRequirement=required
A ->> A: if old password is not empty:<br/>Verify old password
A ->> A: Verify assertion<br/>scope=PASSWORD_CHANGE<br/>UVR as above
```

(I don't know if this renders alright, so feel free to change as you prefer.)

Reply:
@codingllama I'm sorry, I don't understand what "scope=PASSWORD_CHANGE" and "UVR as above" mean in the last line.

Reply:
When you verify an assertion, like in the ChangePassword RPC, you have to say what are the ChallengeExtensions you expect to see. In our case, we would verify using scope=PASSWORD_CHANGE and UVR="required" (UserVerificationRequirement). The fact that we say what extensions we expect to see is what guarantees the "wrong" challenge won't be accepted.

(UVR is just a shortened "UserVerificationRequirement". Feel free to spell it out.)

Reply:
Btw, this some of this is in RFD 155 - https://github.com/gravitational/teleport/blob/master/rfd/0155-scoped-webauthn-credentials.md.

Reply:
@codingllama I think I addressed what you wrote, though I made it a bit more clear. Please take a look. (I also included a reference to RFD 155).

Reply:
Looks good!